cpp/src/arrow/flight/internal.cc

```diff
@@ -45,12 +46,75 @@ namespace flight {
 namespace internal {
 
 const char* kGrpcAuthHeader = "auth-token-bin";
+const char* kGrpcStatusCodeHeader = "x-arrow-status";
+const char* kGrpcStatusMessageHeader = "x-arrow-status-message-bin";
+const char* kGrpcStatusDetailHeader = "x-arrow-status-detail-bin";
+
+/// Try to extract a status from gRPC trailers. If not found, return
+/// Status::OK.
+Status FromGrpcContext(const grpc::ClientContext& ctx) {
+  const std::multimap<grpc::string_ref, grpc::string_ref> trailers =
+      ctx.GetServerTrailingMetadata();
+  const auto code_val = trailers.find(kGrpcStatusCodeHeader);
+  if (code_val == trailers.end()) {
+    return Status::OK();
+  }
+
+  // Bounce through std::string to get a proper null-terminated C string
+  const grpc::string_ref code_ref = (*code_val).second;
+  const auto code_int = std::atoi(std::string(code_ref.data(), code_ref.size()).c_str());
+  StatusCode code = StatusCode::OK;
+  switch (code_int) {
```

Comment:
Is this worth refactoring out into status.h?

Reply:
(this being the int<->statuscode mapping)

Reply:
Hmm... perhaps as a helper function?

Reply:
I've made a helper function, but kept it internal, since there doesn't seem to be a need to serialize statuses outside of Flight.